src/server/journal/journal_slice.cc

```diff
@@ -116,36 +117,63 @@ error_code JournalSlice::Close() {
 }
 #endif
 
+bool JournalSlice::IsLSNInBuffer(LSN lsn) const {
+  DCHECK(ring_buffer_);
+
+  if (ring_buffer_->empty()) {
+    return false;
+  }
+  return (*ring_buffer_)[0].lsn <= lsn && lsn <= ((*ring_buffer_)[ring_buffer_->size() - 1].lsn);
+}
+
+std::string_view JournalSlice::GetEntry(LSN lsn) const {
+  DCHECK(ring_buffer_ && IsLSNInBuffer(lsn));
+  auto start = (*ring_buffer_)[0].lsn;
+  DCHECK((*ring_buffer_)[lsn - start].lsn == lsn);
+  return (*ring_buffer_)[lsn - start].data;
+}
+
 void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
-  // DCHECK(ring_buffer_);
-  if (entry.opcode != Op::NOOP) {
-    lsn_++;
-// TODO: This is preparation for AOC style journaling, currently unused.
-#if 0
-    RingItem item;
-    item.lsn = prev_lsn;
+  DCHECK(ring_buffer_);
+
+  JournalItem dummy;
+  JournalItem* item;
+  if (entry.opcode == Op::NOOP) {
+    item = &dummy;
+    item->lsn = -1;
+    item->opcode = entry.opcode;
+    item->data = "";
+  } else {
+    item = ring_buffer_->GetTail(true);
```

Comment:
1. Please add comment saying, it provide a pointer to next item, overriding the next one if full.
2. I would like to understand how you envision product behavior in this case. Are you saying, there is no backpressure whatsoever, and if this item has not been consumed by a replica, then cest la vie?
3. How replica will know that its stable sync channel was overridden and its missing some entries?